jetty-client/src/main/java/org/eclipse/jetty/client/util/AbstractRequestContent.java

```diff
@@ -0,0 +1,250 @@
+                    {
+                        if (!produceContent(this::processContent))
+                            return;
+                    }
+                    catch (Throwable x)
+                    {
+                        // Fail and loop around to notify the failure.
+                        fail(x);
+                    }
+                }
+                else
+                {
+                    if (!processContent(BufferUtil.EMPTY_BUFFER, false, Callback.NOOP))
+                        return;
+                }
+            }
+        }
+
+        protected abstract boolean produceContent(Producer producer) throws Exception;
+
+        @Override
+        public void fail(Throwable failure)
+        {
+            try (AutoLock ignored = lock.lock())
+            {
+                if (this.failure == null)
+                    this.failure = failure;
+            }
+        }
+
+        private boolean processContent(ByteBuffer content, boolean last, Callback callback)
+        {
+            try (AutoLock ignored = lock.lock())
+            {
+                committed = true;
+                --demand;
+            }
+
+            if (content != null)
+                notifyContent(content, last, callback);
+            else
+                callback.succeeded();
+
+            boolean noDemand;
+            try (AutoLock ignored = lock.lock())
+            {
+                noDemand = demand == 0;
+                if (noDemand)
+                    stalled = true;
+            }
+            if (noDemand)
+            {
+                if (LOG.isDebugEnabled())
+                    LOG.debug("No demand, processing stalled for {}", this);
+                return false;
+            }
+            return true;
+        }
+
+        protected void notifyContent(ByteBuffer buffer, boolean last, Callback callback)
+        {
+            try
+            {
+                if (LOG.isDebugEnabled())
+                    LOG.debug("Notifying content last={} {} for {}", last, BufferUtil.toDetailString(buffer), this);
+                consumer.onContent(buffer, last, callback);
+            }
+            catch (Throwable x)
+            {
+                callback.failed(x);
+                fail(x);
+            }
+        }
+
+        private void notifyFailure(Throwable failure)
+        {
+            try
+            {
+                if (LOG.isDebugEnabled())
+                    LOG.debug("Notifying failure for {}", this, failure);
+                consumer.onFailure(failure);
+            }
+            catch (Exception x)
+            {
+                LOG.trace("Failure while notifying content failure {}", failure, x);
+            }
+        }
+
+        @Override
+        public String toString()
+        {
+            int demand;
+            boolean stalled;
+            try (AutoLock ignored = lock.lock())
+            {
+                demand = this.demand;
+                stalled = this.stalled;
```

Comment:
you are missing `committed` and `submitInitialContent`

Reply:
Will do.